### PR TITLE
Make schema infer work for CREATE AS SELECT

### LINKED ISSUE #47599: Schema inference for INSERT SELECT should also work for CREATE AS SELECT
```
milovidov-desktop :) CREATE TABLE `01902_db`.t (`n` Int8) ENGINE = MergeTree ORDER BY n AS SELECT * FROM generateRandom() LIMIT 10;

CREATE TABLE `01902_db`.t
(
    `n` Int8
)
ENGINE = MergeTree
ORDER BY n AS
SELECT *
FROM generateRandom()
LIMIT 10

Query id: 2003bf93-4643-4dca-8319-5eb562190805


0 rows in set. Elapsed: 0.047 sec. 

Received exception:
Code: 636. DB::Exception: Table function 'generateRandom' was used without structure argument but structure could not be determined automatically. Please, provide structure manually. (CANNOT_EXTRACT_TABLE_STRUCTURE)
```

### PR DESCRIPTION
**Pull request by ucasfl**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Make Schema inference works for `CREATE AS SELECT`. Closes https://github.com/ClickHouse/ClickHouse/issues/47599.
.
Previous PR is #48679, and reverted in #48758.

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
This is an automated comment for commit 845e9847c7a89ed60aedad19cd159bc6afc12334 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49899/845e9847c7a89ed60aedad19cd159bc6afc12334/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by ucasfl** [comment]
It should works well after https://github.com/ClickHouse/ClickHouse/pull/53348.

**Comment by Avogar** [comment]
Performance test is still failing...

**Comment by alexey-milovidov** [comment]
@ucasfl could you please merge with master?

**Comment by ucasfl** [comment]
Still failed...

File: src/Interpreters/InterpreterInsertQuery.cpp [hunk 1]
```diff
@@ -456,7 +456,6 @@ BlockIO InterpreterInsertQuery::execute()
 
                 auto new_context = Context::createCopy(context);
                 new_context->setSettings(new_settings);
-                new_context->setInsertionTable(getContext()->getInsertionTable());
```
Comment (Avogar) [inline src/Interpreters/InterpreterInsertQuery.cpp:459]: Do you recall why this was deleted?

Comment (ucasfl) [reply src/Interpreters/InterpreterInsertQuery.cpp:459]: It's not used I remember.

**Pull request closed without merging**